### ISSUE TITLE
`hs -c "hs.reload()"` spews errors and doesn't reload

### ISSUE THREAD

**Issue by chdiza** [Hammerspoon/hammerspoon#1548]
```bash
$ hs -c "hs.reload()"
error communicating with Hammerspoon: message port was invalidated
error unregistering CLI instance with Hammerspoon: message port invalid
```
And then, HS doesn't actually reload.

**Comment by asmagill** [comment]
Interestingly for me I get the same output, Hammerspoon starts to reload and then crashes.

However, `hs -c "hs.timer.doAfter(1, hs.reload)"` works as expected.

I'll have to look into it.

**Comment by daviehh** [comment]
can replicate with the newest version 0.9.56, it used to work okay before the update.

**Comment by cmsj** [comment]
Looks like the crash is happening in `hs.ipc:sendMessage()` - perhaps because it's trying to send a message using a lua_State object that doesn't exist anymore due to the reload?

**Comment by asmagill** [comment]
The crash is occurring in `ipc_callback`because the external tool `hs` is sending the reload command, then immediately following it with a request for a clean disconnect, but the lua state is in the process of being torn down... checking the state with `lua_status` in `ipc_callback` doesn't detect this (it's still considered valid during teardown), so I'm going to have to think a little before proposing and testing a solution.

In the mean time, if you need to do a reload through the command line tool, I suggest using `hs.timer.doAfter(1, hs.reload)`

**Comment by asmagill** [comment]
Clarification - that's not quite accurate... the first call to `ipc_callback` is to set up the environment for the cli tool, the second is the actual command `hs.reload` and the call to `[skin protectedCallAndTraceback:nresults:]` never returns, which apparently is preventing the ipc module from being properly collected during the destroy process.

I need to look closer at how the original version differs... on the surface, I would expect both versions to have failed this way, so I must be missing something.

**Comment by asmagill** [comment]
@cmsj, you're partially right that ipc_sendmessage is at fault... `hs.reload` does not actually do the reload... it queues it on the runloop and then returns (e.g. `hs.reload() ; t = os.time() ; hs.timer.usleep(10000000) ; print(os.time() - t)` will delay the reload for 10 seconds).  This means that any code which follows the reload command but does not trigger adding something else to the runloop queue will get executed before the reload actually takes place.  In our case it is `ipc_sendmessage` that adds to the runloop.

I have an idea about wrapping `hs.reload` so that `ipc` can detect when it has been triggered that should circumvent this but can't try it out until later tonight.  Long term, though, we need to decide if this "loop hole" in the `hs.reload` command is an area of concern or not... I think the worst that someone can do is crash Hammerspoon like this does, but I'm not positive and will have to consider it more deeply.

**Comment by asmagill** [comment]
This should be addressed with pull #1556. Please reopen if you disagree or continue to have problems.
